### PR TITLE
Fix simclr loss to work with distributed training.

### LINKED ISSUE #258: distributed training  support
When using distributed training the simclr loss should be calculated on all samples across gpus like here:
https://github.com/google-research/simclr/blob/master/objective.py
Is it planned to add this functionality?

### PR DESCRIPTION
**Pull request by yonigottesman**
fix  #258.
Also, while fixing this I noticed the current implementation is wrong even on a single gpu:
The paper states:
"The final loss is computed across all positive pairs, both (i, j) and (j, i)" meaning the loss is computed twice, once (za,zb) and once (zb,za)

To check that this works I tried to run this code on a multi gpu (4) instance:

~~~python
from tensorflow_similarity.losses.simclr import SimCLRLoss
import tensorflow as tf

global_batch_size=512

h1 = tf.random.normal((global_batch_size,256))
h2 = tf.random.normal((global_batch_size,256))

l = SimCLRLoss()(h1,h2)
assert l == tf.reduce_sum(SimCLRLoss(reduction=tf.keras.losses.Reduction.NONE)(h1,h2))/global_batch_size

dataset = tf.data.Dataset.from_tensor_slices((h1,h2)).batch(global_batch_size)
bh1,bh2 = next(iter(dataset))
assert l == SimCLRLoss()(bh1,bh2)

strategy = tf.distribute.MirroredStrategy(devices=["GPU:0", "GPU:1","GPU:2", "GPU:3"])
dist_dataset = strategy.experimental_distribute_dataset(dataset)
with strategy.scope():
    def replica_fn(input):
        h1,h2 = input
        return SimCLRLoss(reduction=tf.keras.losses.Reduction.NONE)(h1,h2)
    x = next(iter(dist_dataset))
    result = strategy.run(replica_fn, args=(x,))

l_distributed = tf.reduce_sum(strategy.reduce("SUM", result, axis=None))/global_batch_size
abs(l-l_distributed)

>>> <tf.Tensor: shape=(), dtype=float32, numpy=9.536743e-07>
~~~
As you see the result is not bit exact (e-7) but this is the exact result compared to original implementation.

Im not sure how to test this behavior because tests are probably not run in a multi gpu instance. maybe the tf.strategy can context can be mocked somehow just for testing?

### CODE REVIEW

**Comment by yonigottesman** [comment]
@owenvallis thanks for clarifying (in the issue thread), indeed the forward pass does the (za,zb) (zb,za). personally I think its better to have the whole loss computation in a single place, its a bit less confusing, it looks exactly like the paper loss and its easier (for me :) ) to reason about the whole averaging over batch size across gpus. 
Its up to you, if you prefer to call it in the forward pass Ill update the loss, if you prefer to call it through the loss ill update the forward pass. 
Im still not sure why the scaling and margin is needed, the original implementations doesn't do it. Did you see better results using it??

**Comment by owenvallis** [comment]
@yonigottesman you raise a good point. The motivation for handling this in the forward pass was to provide support for the different contrastive losses. SimSiam requires 4 inputs, both sets of projector and predictor outputs, while barlow twins only requires a single call using za, zb.

I suppose we could have Simclr and Barlow Twins accept za and zb, and then use the conditional check for Simsiam to handle the 4 input case. I'll take a look at it this weekend.

File: tensorflow_similarity/losses/simclr.py [hunk 1]
```diff
@@ -16,81 +16,122 @@
...
+            masks = tf.one_hot(labels_idx, enlarged_batch_size)
+        else:
+            hidden1_large = hidden1
+            hidden2_large = hidden2
+            labels = tf.one_hot(tf.range(batch_size), batch_size * 2)
+            masks = tf.one_hot(tf.range(batch_size), batch_size)
+
+        # compute pairwise
+        ab = tf.matmul(hidden1, hidden2_large, transpose_b=True) / self.temperature
+        aa = tf.matmul(hidden1, hidden1_large, transpose_b=True) / self.temperature
```
Comment (owenvallis) [inline tensorflow_similarity/losses/simclr.py:64]: Should this be hidden2?

`tf.matmul(hidden2, hidden1_large, transpose_b=True)`

**Review by owenvallis — CHANGES_REQUESTED** [review]
Looks good, and thanks again for adding support for this! 

I left one comment where I think we want hidden2 instead of hidden1, but otherwise looks good.

File: tensorflow_similarity/losses/simclr.py (hunk 1 above)
Comment (yonigottesman) [reply tensorflow_similarity/losses/simclr.py:64]: no. aa holds the cosine similarity between all vectors in hidden1 and itself. later we make the diag -inf to remove the dist(e_i,e_i) from the denominator in thesum :
image
Its like it was before:
`aa = tf.matmul(za, za, transpose_b=True)`

Comment (owenvallis) [reply tensorflow_similarity/losses/simclr.py:64]: You're totally right, I'd forgotten about that. Thanks for clarifying and for the PR.

**Review by owenvallis — APPROVED** [review]
